utilities/generate_messages_catalog.py

```diff
@@ -0,0 +1,326 @@
+                    else:
+                        level, args_offset = handler(item)
+
+                    msgCodeArg = item.args[0 + args_offset]  # str or tuple
+                    if isinstance(msgCodeArg,ast.Str):
+                        msgCodes = (msgCodeArg.s,)
+                    else:
+                        if any(isinstance(element, (ast.Call, ast.Name))
+                               for element in ast.walk(msgCodeArg)):
+                            msgCodes = ("(dynamic)",)
+                        else:
+                            msgCodes = [
+                                element.s for element in ast.walk(msgCodeArg)
+                                if isinstance(element, ast.Str)
+                            ]
+                    msgArg = item.args[1 + args_offset]
+                    if isinstance(msgArg, ast.Str):
+                        msg = msgArg.s
+                    elif ((isinstance(msgArg, ast.Call) and
+                           getattr(msgArg.func, "id", '') == '_')):
+                        msg = msgArg.args[0].s
+                    elif ((any(isinstance(element, (ast.Call,ast.Name))
+                           for element in ast.walk(msgArg)))):
+                        msg = "(dynamic)"
+                    else:
+                        continue # not sure what to report
+                    keywords = []
+                    for keyword in item.keywords:
+                        if keyword.arg == 'modelObject':
+                            pass
+                        elif keyword.arg == 'messageCodes':
+                            msgCodeArg = keyword.value
+                            if ((any(isinstance(element, (ast.Call, ast.Name))
+                                 for element in ast.walk(msgCodeArg)))):
+                                pass  # dynamic
+                            else:
+                                msgCodes = [
+                                    element.s
+                                    for element in ast.walk(msgCodeArg)
+                                    if isinstance(element, ast.Str)
+                                ]
+                        else:
+                            keywords.append(keyword.arg)
+                    for msgCode in msgCodes:
+                        id_messages.append(
+                            {
+                                'message_code': msgCode,
+                                'message': entityEncode(msg),
+                                'level': level,
+                                'keyword_arguments': entityEncode(
+                                    " ".join(keywords)
+                                ),
+                                'reference_filename': refFilename,
+                                'line_number': item.lineno
+                            }
+                        )
+                except (AttributeError, IndexError):
+                    pass
```

Comment:
I actually need to ask Herm about the index one, since the attribute error is actually handled by the extracted function. It is on my list of things to tackle in the next phase.